<a href="https://colab.research.google.com/github/santiagoramosg/Fundamentos-de-optimizacion-lineal/blob/main/Taller_03_Branch_and_Bound.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## «*Dices que cada nueva mañana nos trae mil rosas sí, pero, ¿dónde están los pétalos de la rosa de ayer?*»
### [Omar Khayyam](https://es.wikipedia.org/wiki/Omar_Jayam)

# Taller 03 Branch and Bound



# Branch and Bound

Ejemplo de la solución  paso a paso de un problema de programación entera utilizando la heurística Branch and Bound.

## Problema inicial: problema00

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [1]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones00 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob00 = cvx.Problem(obj, restricciones00)

# Obtener el valor óptimo
prob00.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob00.status)
print("Valor máximo:", "%.2f"% prob00.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.25
Valor óptimo de x: 3.75
Valor óptimo de y: 2.25


## subproblema01

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45
   
   **x <= 3**

   x   > = 0      

   y   >=  0

**x, y enteros**


In [5]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones01 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
# Ramificación del problema00
x <= 3,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob01 = cvx.Problem(obj, restricciones01)

# Obtener el valor óptimo
prob01.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob01.status)
print("Valor máximo:", "%.2f"% prob01.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 39.00
Valor óptimo de x: 3.00
Valor óptimo de y: 3.00


## subproblema02

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

   y   >=  0

**x, y enteros**


In [2]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones02 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
 # Ramificación del problema00
x >= 4,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob02 = cvx.Problem(obj, restricciones02)

# Obtener el valor óptimo
prob02.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob02.status)
print("Valor máximo:", "%.2f"% prob02.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.80


## subproblema03

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

  **y >= 2**

**x, y enteros**


In [10]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones03 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
# Ramificación del problema00
x >= 4,

# Ramificación del problema02
y >= 2

# Restricciones de no negatividad

]

# Configurar el problema
prob03 = cvx.Problem(obj, restricciones03)

# Obtener el valor óptimo
prob03.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob03.status)
print("Valor máximo:", "%.2f"% prob03.value)
try:
  print ("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value)
except:
  print ("Valor óptimo de x:", x.value)
  print("Valor óptimo de y:", y.value)

# Mostrar los valores óptimos de las variables de decisión



Estado de la solución: infeasible
Valor máximo: -inf
Valor óptimo de x: None
Valor óptimo de y: None


## subproblema04

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [7]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones04 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
# Ramificación del problema00
x >= 4,

# Ramificación del problema02
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob04 = cvx.Problem(obj, restricciones04)

# Obtener el valor óptimo
prob04.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob04.status)
print("Valor máximo:", "%.2f"% prob04.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.56
Valor óptimo de x: 4.44
Valor óptimo de y: 1.00


## subproblema05

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 5**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [11]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones05 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
# Ramificación del problema04
x >= 5,

# Ramificación del problema02
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob05 = cvx.Problem(obj, restricciones05)

# Obtener el valor óptimo
prob05.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob05.status)
print("Valor máximo:", "%.2f"% prob05.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00


## subproblema06

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **x <= 4**
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [12]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear la variable a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones06 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
# Ramificación del problema00
x >= 4,

# Ramificación del problema04
x <=4,

# Ramificación del problema02
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob06 = cvx.Problem(obj, restricciones06)

# Obtener el valor óptimo
prob06.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob06.status)
print("Valor máximo:", "%.2f"% prob06.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 37.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.00


# Resumen del Branch and Bound
## Problema inicial: problema00
Restricciones: x>=0; y>=0     

Resultados: valor óptimo: 41.25; x.value = 3.75; y.value = 2.25

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0     

Resultados: valor óptimo: 39.00; **x.value = 3.00**; **y.value = 3.00**

## subproblema02
Restricciones: **x>=4**; y>=0     

Resultados: valor óptimo: 41.00; x.value = 4.00; y.value = 1.80

## subproblema03
Restricciones: **x>=4**; **y >=2**     

Resultados: valor óptimo: infeasible; x.value = ???; y.value = ???

## subproblema04
Restricciones: **x>=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 40.56; x.value = 4.44; y.value = 1.00

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: valor óptimo: 40.00; **x.value = 5.00; y.value = 0.00**

## subproblema06
Restricciones: **x>=4**; **x<=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 37.00; **x.value = 4.0; y.value = 1.0**

# Análisis de los resultados del Branch and Bound

Los subproblemas 01, 05 y 06 tienen una solución óptima factible (soluciones enteras).

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0

Resultados: **valor óptimo: 39.00; x.value = 3.00; y.value = 3.00**

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: **valor óptimo: 40.00; x.value = 5.0; y.value = 0.0**

## subproblema06
Restricciones: x>=4; x<=4; y>=0; y <=1

Resultados: **valor óptimo: 37.00; x.value = 4.0; y.value = 1.0**

Se selecciona como solución optima la del subproblema 05, ya que satisface las restricciones de enteros y genera el mayor valor óptimo.


# Problema inicial: problema00 resuelto directamente por CVXPY

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [13]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar y establecerlas como enteros
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones_int = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob_int = cvx.Problem(obj, restricciones_int)

# Obtener el valor óptimo
prob_int.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob_int.status)
print("Valor máximo:", "%.2f"% prob_int.value)

# Mostrar los valores óptimos de las variables de decisión
print("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value, "\n")

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00 



# Ejercicios
Resuelva con el método **Branch and Bound**, utilizando CVXPY, cada uno de los siguientes ejercicios.

***Agregue todas las casillas de texto y de código que requiera***.

## Ejercicio 00

Maximizar  Z: 3x + 2y

   sujeto a:

   2x + y <= 10

   x + 3y <= 12

  **x, y son enteros no negativos**
  






In [36]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


# Importar la librería que se va a utilizar
import cvxpy as cvx


def bab_optimal_int():

  # Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable(integer=True)
  y = cvx.Variable(integer=True)
  # Definir la función objetivo para maximizar el ingreso
  obj = cvx.Maximize(3*x + 2*y)
  # Establecer las restricciones
  restricciones= [
  # Primera restricción
  2*x + y <= 10,
  # Segunda restricción
  x + 3*y <= 12,
  # Restricciones de no negatividad
  x >= 0,
  y >= 0
  ]
  # Configurar el problema
  prob = cvx.Problem(obj, restricciones)
  # Obtener el valor óptimo
  prob.solve()
  # Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)
  # Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")



#!caso base !#
def bab():

  # Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable()
  y = cvx.Variable()
  # Definir la función objetivo para maximizar el ingreso
  obj = cvx.Maximize(3*x + 2*y)
  # Establecer las restricciones
  restricciones= [
  # Primera restricción
  2*x + y <= 10,
  # Segunda restricción
  x + 3*y <= 12,
  # Restricciones de no negatividad
  x >= 0,
  y >= 0
  ]
  # Configurar el problema
  prob = cvx.Problem(obj, restricciones)
  # Obtener el valor óptimo
  prob.solve()
  # Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)
  # Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")


#!sub caso 1! x<=3 #
def bab_1():

  # Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable()
  y = cvx.Variable()
  # Definir la función objetivo para maximizar el ingreso
  obj = cvx.Maximize(3*x + 2*y)
  # Establecer las restricciones
  restricciones= [
  # Primera restricción
  2*x + y <= 10,
  # Segunda restricción
  x + 3*y <= 12,

  # Restriccion sub problema 1
  x <= 3,

  # Restricciones de no negatividad
  x >= 0,
  y >= 0
  ]
  # Configurar el problema
  prob = cvx.Problem(obj, restricciones)
  # Obtener el valor óptimo
  prob.solve()
  # Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)
  # Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")


#!sub caso 2! x>=4#
def bab_2():

  # Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable()
  y = cvx.Variable()
  # Definir la función objetivo para maximizar el ingreso
  obj = cvx.Maximize(3*x + 2*y)
  # Establecer las restricciones
  restricciones= [
  # Primera restricción
  2*x + y <= 10,
  # Segunda restricción
  x + 3*y <= 12,

  #restricion sub problema 2
  x >= 4,

  # Restricciones de no negatividad
  x >= 0,
  y >= 0
  ]
  # Configurar el problema
  prob = cvx.Problem(obj, restricciones)
  # Obtener el valor óptimo
  prob.solve()
  # Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)
  # Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")


x= int(input("ingrese prueba\n"+
             "0 caso optimo\n"+
             "1 caso base\n"+
             "2 sub problema 1 \n"+
             "3 sub problema 2 \n"))

match x:
  case 0:
    bab_optimal_int()
  case 1:
    bab()
  case 2:
    bab_1()
  case 3:
    bab_2()

"""Importar la librería que se va a utilizar
Crear las variables a optimizar y establecerlas como enteros
Definir la función objetivo para maximizar el ingreso
Establecer las restricciones
Configurar el problema
Obtener el valor óptimo
Mostrar el estado de la solución y el valor óptimo de la función objetivo
Mostrar los valores óptimos de las variables de decisión"""

ingrese prueba
0 caso optimo
1 caso base
2 sub problema 1 
3 sub problema 2 
3
Estado de la solución: optimal
Valor máximo: 16.00
Valor óptimo de x: 4.00
Valor óptimo de y: 2.00 



'Importar la librería que se va a utilizar\nCrear las variables a optimizar y establecerlas como enteros\nDefinir la función objetivo para maximizar el ingreso\nEstablecer las restricciones\nConfigurar el problema\nObtener el valor óptimo\nMostrar el estado de la solución y el valor óptimo de la función objetivo\nMostrar los valores óptimos de las variables de decisión'

##resultados BAB

BASE
X=3.60, Y=2.80 Maximo =16.40

SUB PROBLEMA 1
X=3, Y=3, Maximo=15

SUB PROBLEMA 2
X=4, Y=2, Maximo=16


OPTIMAL
x=4, Y=2 Valor Maximo=16

## Ejercicio 01

Minimizar  Z: 2x + 3y

   sujeto a:

   3x + y >= 6

   2x + 2y >= 8

   x >= 0

   y >= 0



In [42]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


#Importar la librería que se va a utilizar
import cvxpy as cvx


#!caso optimo!#
def bab_optimal():
  #Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable(integer=True)
  y = cvx.Variable(integer=True)
  #Definir la función objetivo para maximizar el ingreso
  obj = cvx.Minimize(2*x + 3*y)

  #Establecer las restricciones
  restricciones=[
  #primera restriccion
  3*x+y>=6,
  #segunda restricion
  2*x+2*y>=8,
  #restricciones de no negatividad
  x>=0,
  y>=0
  ]

  #Configurar el problema
  prob=cvx.Problem(obj,restricciones)
  #Obtener el valor óptimo
  prob.solve()
  #Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)

  #Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")

#!caso base!#
def bab():
  #Crear las variables a optimizar y establecerlas como enteros
  x = cvx.Variable()
  y = cvx.Variable()
  #Definir la función objetivo para maximizar el ingreso
  obj = cvx.Minimize(2*x + 3*y)

  #Establecer las restricciones
  restricciones=[
  #primera restriccion
  3*x+y>=6,
  #segunda restricion
  2*x+2*y>=8,
  #restricciones de no negatividad
  x>=0,
  y>=0
  ]

  #Configurar el problema
  prob=cvx.Problem(obj,restricciones)
  #Obtener el valor óptimo
  prob.solve()
  #Mostrar el estado de la solución y el valor óptimo de la función objetivo
  print("Estado de la solución:", prob.status)
  print("Valor máximo:", "%.2f"% prob.value)

  #Mostrar los valores óptimos de las variables de decisión
  print("Valor óptimo de x:", "%.2f"% x.value)
  print("Valor óptimo de y:", "%.2f"% y.value, "\n")


x= int(input("ingrese prueba\n"+
             "0 caso optimo\n"+
             "1 caso base\n"))

match x:
  case 0:
    bab_optimal()
  case 1:
    bab()

"""#Importar la librería que se va a utilizar
#Crear las variables a optimizar y establecerlas como enteros
#Definir la función objetivo para maximizar el ingreso
#Establecer las restricciones
#Configurar el problema
#Obtener el valor óptimo
#Mostrar el estado de la solución y el valor óptimo de la función objetivo
#Mostrar los valores óptimos de las variables de decisión"""

ingrese prueba
0 caso optimo
1 caso base
1
Estado de la solución: optimal
Valor máximo: 8.00
Valor óptimo de x: 4.00
Valor óptimo de y: -0.00 



'#Importar la librería que se va a utilizar\n#Crear las variables a optimizar y establecerlas como enteros\n#Definir la función objetivo para maximizar el ingreso\n#Establecer las restricciones\n#Configurar el problema\n#Obtener el valor óptimo\n#Mostrar el estado de la solución y el valor óptimo de la función objetivo\n#Mostrar los valores óptimos de las variables de decisión'

##RESULTADOS BAB

BASE
X=4 Y=-0 MAXIMO=8

OPTIMAL
X=4, Y=0, MAXIMO=8

## Ejercicio 02

Maximizar  Z: 4x + 3y

   sujeto a:

   3x + 2y <= 15

   2x + 4y <= 12

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 03

Minimizar  Z: 4x + 2y

   sujeto a:

   x + 3y >= 9

   2x + y >= 8

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 04

Maximizar  Z: 5x + 4y

   sujeto a:

   x + y <= 8

   2x + 3y <= 15

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 05

Minimizar  Z: 3x + 4y

   sujeto a:

   x + 2y >= 5

   2x + y >= 7

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 06

Maximizar  Z: 2x + 5y

   sujeto a:

   x + 2y <= 6

   3x + y <= 9

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 07

Minimizar  Z: x + 2y

   sujeto a:

   2x + y >= 10

   x + 3y >= 12

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 08

Maximizar  Z: x + 3y

   sujeto a:

   x + y <= 5

   2x + 2y <= 10

   x + 2y <= 8

  **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera


## Ejercicio 09

Minimizar  Z: 2x + y

   sujeto a:

   x + y >= 6

   2x + 3y >= 12

   x + 2y >= 8

   **x, y son enteros no negativos**

In [ ]:
# Resuelva a partir de esta celda el ejercicio.
# Agregue todas las casillas de texto y de código que requiera
